### PR TITLE
Layout CME and SGPE rings in HOMER image

### PR DESCRIPTION
**Pull request by SergiiDmytruk**

### CODE REVIEW

File: src/soc/ibm/power9/homer.c [hunk 1]
```diff
@@ -17,6 +18,211 @@
 
 #include <lib.h>
 
+#define CACHE0_CHIPLET_ID 0x10
+#define CORE0_CHIPLET_ID  0x20
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: We have `EP00_CHIPLET_ID` and `EC00_CHIPLET_ID` in scom.h, use them instead.

File: src/soc/ibm/power9/homer.c [hunk 2]
```diff
@@ -17,6 +18,211 @@
...
+
+	/* pstate-volt threshold slopes */
+	int16_t PsVDMThreshSlopes[VPD_NUM_SLOPES_REGION][NUM_THRESHOLD_POINTS];
+
+	/* Jump-value operating points */
+	uint8_t jump_value_set[NUM_OP_POINTS][NUM_JUMP_VALUES];
+
+	/* Jump-value slopes */
+	int16_t PsVDMJumpSlopes[VPD_NUM_SLOPES_REGION][NUM_JUMP_VALUES];
+};
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c:219]: Most of the above (if not all) is in files added by `pstates` branch, so either a rebase or another fix after merging both into `homer` will be needed.

File: src/soc/ibm/power9/homer.c [hunk 3]
```diff
@@ -981,14 +1510,20 @@ void build_homer_image(void *homer_bar)
 	build_pgpe(homer, (struct xip_pgpe_header *)(homer_bar + hw->pgpe.offset),
 	           dd);
 
-	// "test" of tor_fetch_and_insert_vpd_rings()
-	getPpeScanRings(hw, dd);
+	ring_variant = (dd < 23 ? RV_BASE : RV_RL4);
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]:
```suggestion
	ring_variant = (dd < 0x23 ? RV_BASE : RV_RL4);
```

File: src/soc/ibm/power9/homer.c [hunk 4]
```diff
@@ -908,22 +1109,350 @@ static void getPpeScanRings(struct xip_hw_header *hw, uint8_t dd)
...
+		/* Skip non-functional quads */
+		if (!IS_EQ_FUNCTIONAL(quad, cores))
+		    continue;
+
+		for (i = 0; i < sizeof(ring_ids) / sizeof(ring_ids[0]); ++i) {
+			const enum ring_id id = ring_ids[0];
+
+			uint32_t ring_size = MAX_RING_BUF_SIZE;
+
+			uint8_t chiplet_id = CACHE0_CHIPLET_ID + quad*2;
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Please either change that variable name or add a comment that this is **not** an SCOM chiplet ID, or both. I know it is called like that in Hostboot, but it is misleading - there are only 6 cache chiplets.

File: src/soc/ibm/power9/homer.c [hunk 5]
```diff
@@ -908,22 +1109,350 @@ static void getPpeScanRings(struct xip_hw_header *hw, uint8_t dd)
...
+			if (i != 0 && (i - 1) % 2 == 1)
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Simpler:
```suggestion
			if (i != 0 && (i % 2) == 0)
```
Sidenote: Hostboot has an original idea of odd/even numbers :slightly_smiling_face:

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: src/soc/ibm/power9/homer.c (hunk 2 above)
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/homer.c:219]: I saw those files, but thought they were added as-is temporarily.

**Review by krystian-hebel — APPROVED** [review]
Code looks good and gives exactly the same result as Hostboot. The only thing I'm concerned about is `ALIGN_UP(payload - start, 8)` when `start` isn't already aligned, but the logic is identical to Hostboot so let's leave it as it is.

File: src/soc/ibm/power9/homer.c (hunk 2 above)
Comment (krystian-hebel) [reply src/soc/ibm/power9/homer.c:219]: That was my initial idea, but Pstate's code uses most of those anyway so I think adding whole files (and not taking the responsibility for misleading field names :slightly_smiling_face: ) will be better.

**Pull request merged**